<a href="https://colab.research.google.com/github/soustan/Internship-Data-Analitic/blob/main/Internship-part-4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [27]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [28]:
import pandas as pd
import numpy as np
import re

In [ ]:
#Load the region roles
from google.colab import files
upload = files.upload()

In [31]:
reg_rol = pd.read_excel('/region_roles_26_05_22.xlsx')

In [ ]:
reg_rol.head(10)

In [ ]:
#reg_rol['salary'].str.split('\n')[0]

nan

In [ ]:
#new_reg_rol = reg_rol['salary'].str.split(',',expand=True)
#new_reg_rol

In [33]:
print(reg_rol['salary'][1])
print(reg_rol['salary'][1].replace("{", '').replace("}", '').split(',')) # превращаем из словаря в список
print(reg_rol['salary'][1].replace("{", '').replace("}", '').split(',')[0][8:]) 
#в списке берем 1-эй элемент
def salary_from(salary):
  if pd.isna(salary):
   clean_salary = np.nan
  else:
   clean_salary = salary.replace("{", '').replace("}", '').split(',')[0][8:]
   if clean_salary == 'None':
     clean_salary = np.nan
  return clean_salary
reg_rol['salary_from'] = reg_rol.salary.apply(salary_from)
reg_rol['salary_from'] = pd.to_numeric(reg_rol['salary_from'])
reg_rol['salary_from'] = reg_rol['salary_from'].astype('Int64')
reg_rol['salary_from'].describe()
reg_rol.info()

{'from': 25000, 'to': 50000, 'currency': 'RUR', 'gross': True}
["'from': 25000", " 'to': 50000", " 'currency': 'RUR'", " 'gross': True"]
25000
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69656 entries, 0 to 69655
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                69656 non-null  int64  
 1   id                        69656 non-null  int64  
 2   premium                   69656 non-null  bool   
 3   name                      69656 non-null  object 
 4   department                5263 non-null   object 
 5   has_test                  69656 non-null  bool   
 6   response_letter_required  69656 non-null  bool   
 7   area                      69656 non-null  object 
 8   salary                    37590 non-null  object 
 9   type                      69656 non-null  object 
 10  address                   40318 non-null  object 
 11  response_url              1 

In [ ]:
print(reg_rol['salary'][1])
print(reg_rol['salary'][1].replace("{", '').replace("}", '').split(',')) # превращаем из словаря в список
print(reg_rol['salary'][1].replace("{", '').replace("}", '').split(',')[1][8:]) 
#в списке берем 2-эй элемент
def salary_to(salary):
  if pd.isna(salary):
    clean_salary = np.nan
  else:
      clean_salary = salary.replace("{", '').replace("}", '').split(',')[1][8:]
      if clean_salary == 'None':
        clean_salary = np.nan
        return clean_salary
reg_rol['salary_to'] = reg_rol.salary.apply(salary_to)
reg_rol['salary_to'] = pd.to_numeric(reg_rol['salary_to'])
reg_rol['salary_to'] = reg_rol['salary_to'].astype('Int64')
reg_rol['salary_to'].describe()
reg_rol.info()

In [52]:
print(reg_rol['salary'][1])
print(reg_rol['salary'][1].replace("{", '').replace("}", '').split(',')) # превращаем из словаря в список
print(reg_rol['salary'][1].replace("{", '').replace("}", '').split(',')[2][8:]) 
#в списке берем 3-эй элемент
def salary_currency(salary):
  if pd.isna(salary):
    clean_salary = np.nan
  else:
      clean_salary = salary.replace("{", '').replace("}", '').split(',')[2][8:]
      if clean_salary == 'None':
        clean_salary = np.nan
        return clean_salary
reg_rol['salary_currency'] = reg_rol.salary.apply(salary_currency)
reg_rol['salary_currency'] = pd.to_numeric(reg_rol['salary_currency'])
reg_rol['salary_currency'] = reg_rol['salary_currency'].astype('Int64')
reg_rol['salary_currency'].describe()
reg_rol.info()

{'from': 25000, 'to': 50000, 'currency': 'RUR', 'gross': True}
["'from': 25000", " 'to': 50000", " 'currency': 'RUR'", " 'gross': True"]
cy': 'RUR'
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69656 entries, 0 to 69655
Data columns (total 37 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                69656 non-null  int64  
 1   id                        69656 non-null  int64  
 2   premium                   69656 non-null  bool   
 3   name                      69656 non-null  object 
 4   department                5263 non-null   object 
 5   has_test                  69656 non-null  bool   
 6   response_letter_required  69656 non-null  bool   
 7   area                      69656 non-null  object 
 8   salary                    37590 non-null  object 
 9   type                      69656 non-null  object 
 10  address                   40318 non-null  object 
 11  response_url           

In [53]:
print(reg_rol['salary'][1])
print(reg_rol['salary'][1].replace("{", '').replace("}", '').split(',')) # превращаем из словаря в список
print(reg_rol['salary'][1].replace("{", '').replace("}", '').split(',')[3][8:]) 
#в списке берем 4-эй элемент
def salary_gross(salary):
  if pd.isna(salary):
    clean_salary = np.nan
  else:
      clean_salary = salary.replace("{", '').replace("}", '').split(',')[2][8:]
      if clean_salary == 'None':
        clean_salary = np.nan
        return clean_salary
reg_rol['salary_gross'] = reg_rol.salary.apply(salary_gross)
reg_rol['salary_gross'] = pd.to_numeric(reg_rol['salary_gross'])
reg_rol['salary_gross'] = reg_rol['salary_gross'].astype('Int64')
reg_rol['salary_gross'].describe()
reg_rol.info()

{'from': 25000, 'to': 50000, 'currency': 'RUR', 'gross': True}
["'from': 25000", " 'to': 50000", " 'currency': 'RUR'", " 'gross': True"]
: True
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69656 entries, 0 to 69655
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                69656 non-null  int64  
 1   id                        69656 non-null  int64  
 2   premium                   69656 non-null  bool   
 3   name                      69656 non-null  object 
 4   department                5263 non-null   object 
 5   has_test                  69656 non-null  bool   
 6   response_letter_required  69656 non-null  bool   
 7   area                      69656 non-null  object 
 8   salary                    37590 non-null  object 
 9   type                      69656 non-null  object 
 10  address                   40318 non-null  object 
 11  response_url              1

In [ ]:
#Чтение входного файла с данными
%%time
new_reg_rol = pd.read_excel(r'region_roles_26_05_22.xlsx')
#Пишем регулярные выражения
#Шаблоны
nrr1 = r'(?<=\d),(?=\d)' # ищем запятую между двумя цифрами
nrr2 = r'(?<=,) ' # ищем пробел после запятой
nrr3 = r'(?<=\d)\.(?=\d)' # ищем точку между двумя цифрами
#Пишем функции, которые применяют регулярные выражения
def f(a):
    try:
        a = re.sub(r'{|}','', a) #Убираем символы "{",  "|" и "}"
        a = re.sub(nrr1, '', a) #Убираем пробелы после запятых
        a = re.sub(nrr2, '.', a) #Заменяем запятые в десятичных числах на точки (432,63 -> 432.63)
        a = a.split(',') #Делим строчку на элементы между запятыми
        return {key: value for (key, value) in [i.split('=') for i in a]} #формируем словарь, где 'столбец': 'его значение' 
    except:
        return np.nan
def f_2(a):
    try:
        a = re.sub(nrr3, ',', a) #Заменяем обратно запятую в десятичных числах на точку (чтобы потом Excel понимал)
        return a
    except:
        return np.nan
%%time
new_reg_rol['dictionary'] = new_reg_rol['salary'].astype('str').apply(f)
%%time
#Словарь преобразуем в отдельные столбцы
j_2 = new_reg_rol['dictionary'].apply(pd.Series)
j_2 = j_2.applymap(f_2)
#Заменяем 'NaN' на пустую ячейку и соединяем все в одну таблицу
j_2.replace({'NaN': np.nan}, inplace=True)
Output = new_reg_rol.join(j_2, lsuffix='_left', rsuffix='_right')
#Запись результата в Excel
writer = pd.ExcelWriter('dataset_region_roles_26_05_22.xlsx')
Output.to_excel(writer, sheet_name = '1', index=False)
writer.save()

UsageError: Line magic function `%%time` not found.


In [ ]:
new_reg_rol.columns = ['From', 'to', 'Currency']
#splited_items = [a.split('=') for a in new_reg_rol[0]]
#new_reg_rol

In [ ]:
new_reg_rol = reg_rol['salary'].str.split(',').\
apply(lambda x: {item[0]: item[1] for item in [a.split('=') for a in x]}).\
apply(pd.Series)

In [ ]:
reg_rol['salary_1'] = reg_rol['salary'].map(lambda x:x. split(',')[0])
reg_rol['salary_2'] = reg_rol['salary'].map(lambda x:x. split(',')[1])
reg_rol['salary_3'] = reg_rol['salary'].map(lambda x:x. split(',')[2])

In [ ]:
reg_rol

In [ ]:
reg_rol.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69656 entries, 0 to 69655
Data columns (total 34 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                69656 non-null  int64  
 1   id                        69656 non-null  int64  
 2   premium                   69656 non-null  bool   
 3   name                      69656 non-null  object 
 4   department                5263 non-null   object 
 5   has_test                  69656 non-null  bool   
 6   response_letter_required  69656 non-null  bool   
 7   area                      69656 non-null  object 
 8   salary                    37590 non-null  object 
 9   type                      69656 non-null  object 
 10  address                   40318 non-null  object 
 11  response_url              1 non-null      object 
 12  sort_point_distance       0 non-null      float64
 13  published_at              69656 non-null  object 
 14  create

В ходе изучения Датасет, удалось оценить его размер, который равен 15.7 MB. Количество строк в датасет - 69656. Количество колонок(столбцов) - 34. В результате чего, стало понятным, что:
Количество булевых строк равно 5, Строк с плавающей запятой равно 1, строк с целым числом равно 2, строк с типом данных object равно 26. Столбцы, с именами id, premium, name, department, type, address и другие, имеют пустые или нулевые значения. В результате, из данных подлежащих анализу. имеются значения: Целые числа, значения отличные от нуля, а так же значения типа object, столбец salary.

In [ ]:
reg_rol = reg_rol.astype'salary': np.int})

In [ ]:
#Load the result roles
from google.colab import files
upload = files.upload()

Saving result roles (16.06.2022).xlsx to result roles (16.06.2022).xlsx


In [ ]:
res_rol = pd.read_excel('result_roles_16_06_2022.xlsx')

In [ ]:
res_rol.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76382 entries, 0 to 76381
Data columns (total 51 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Unnamed: 0                76382 non-null  int64 
 1   id                        76382 non-null  int64 
 2   premium                   76382 non-null  bool  
 3   name                      76382 non-null  object
 4   department                4902 non-null   object
 5   has_test                  76380 non-null  object
 6   response_letter_required  76382 non-null  object
 7   area                      76381 non-null  object
 8   salary                    40874 non-null  object
 9   type                      76379 non-null  object
 10  published_at              76380 non-null  object
 11  created_at                76381 non-null  object
 12  archived                  76380 non-null  object
 13  insider_interview         570 non-null    object
 14  url                   